src/queues/EventRedactionQueue.ts

```diff
@@ -0,0 +1,139 @@
+        await redactUserMessagesIn(client, this.userId, [this.roomId]);
+    }
+
+    public redactionEqual(redaction: QueuedRedaction): boolean {
+        if (redaction instanceof RedactUserInRoom) {
+            return redaction.userId === this.userId && redaction.roomId === this.roomId;
+        } else {
+            return false;
+        }
+    }
+}
+/**
+ * This is a queue for events so that other protections can happen first (e.g. applying room bans to every room).
+ */
+export class EventRedactionQueue {
+    private toRedact: Map<string, QueuedRedaction[]> = new Map<string, QueuedRedaction[]>();
+
+    /**
+     * Test whether the redaction is already present in the queue.
+     * @param redaction a QueuedRedaction.
+     * @returns True if the queue already has the redaction, false otherwise.
+     */
+    public has(redaction: QueuedRedaction): boolean {
+        return !!this.toRedact.get(redaction.roomId)?.find(r => r.redactionEqual(redaction));
+    }
+
+    /**
+     * Adds a QueuedRedaction to the queue to be processed when process is called.
+     * @param redaction A QueuedRedaction to await processing
+     * @returns A boolean that is true if the redaction was added to the queue and false if it was duplicated.
+     */
+    public add(redaction: QueuedRedaction): boolean {
+        if (this.has(redaction)) {
+            return false;
+        } else {
+            let entry = this.toRedact.get(redaction.roomId);
+            if (entry) {
+                entry.push(redaction);
+            } else {
+                this.toRedact.set(redaction.roomId, [redaction]);
+            }return true;
+        }
+    }
+
+    /**
+     * Process the redaction queue, carrying out the action of each QueuedRedaction in sequence.
+     * @param client The matrix client to use for processing redactions.
+     * @param limitToRoomId If the roomId is provided, only redactions for that room will be processed.
+     * @returns A description of any errors encountered by each QueuedRedaction that was processed.
+     */
+    public async process(client: MatrixClient, limitToRoomId?: string): Promise<RoomUpdateError[]> {
+        const errors: RoomUpdateError[] = [];
+        const redact = async (currentBatch: QueuedRedaction[]) => {
+            for (const redaction of currentBatch) {
+                try {
+                    await redaction.redact(client);
+                } catch (e) {
+                    let roomError: RoomUpdateError;
+                    if (e.roomId && e.errorMessage && e.errorKind) {
+                        roomError = e;
+                    } else {
+                        const message = e.message || (e.body ? e.body.error : '<no message>');
+                        roomError = {
+                            roomId: redaction.roomId,
+                            errorMessage: message,
+                            errorKind: ERROR_KIND_FATAL,
```

Comment:
I haven't checked, what does `ERROR_KIND_FATAL` mean?

Reply:
It is the only alternative to `ERROR_KIND_PERMISSION`, from what I could tell, it is used by mjolnir when displaying errors for a command to only show one error of one kind in the management room to prevent spamming it. https://github.com/matrix-org/mjolnir/blob/main/src/ErrorCache.ts#L38 . We could decide that this is bad when commands are doing lots of things though as it is hard to tell what part of them went wrong.

Reply:
Feels like something that we may need to look at at a later stage. No emergency, though.